Add contribution guideline (#1162)
kanisterio/kanister: CONTRIBUTING.md

# Contributing to Kanister

Submitting issues
If you find a bug or have a feature request, please submit an issue at https://github.com/kanisterio/kanister/issues

> 

Reviewer: ?

# Contributing to Kanister

## How Can I Contribute?
  ### * Reporting Bugs

>   Before creating bug reports, please check the list of issues as you might find out that you don't need to create one. Also when creating a bug report, please include as many details as possible as this information may help us resolve the issue faster. To report a bug please submit an issue at https://github.com/kanisterio/kanister/issues

Reviewer: ```suggestion
  Before creating bug reports, please check the list of issues as you might find out that a similar issue is already created and you don't need to create one. Also when creating a bug report, please include as many details as possible as this information may help us resolve the issue faster.
```

  ### * Suggesting Enhancements

>   You can suggest minor improvements to existing functionality and or a completely new feature. For that  we recommend creating an issue at https://github.com/kanisterio/kanister/issues and labeling it design outlining the feature that you wish to build, and describing how it will be implemented. This gives a chance for review to happen early and ensures no wasted effort occurs.

Reviewer: ```suggestion
  You can suggest minor improvements to existing functionality and or a completely new feature. For that  we recommend creating an issue at https://github.com/kanisterio/kanister/issues or reaching out to us at our slack workspace.
```

Reviewer: what did you mean by `and labeling it design`?

  ### * Your first code contribution
  * You can start by looking through `good-first-issue` issues. We follow the [Github Pull Request Model](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) for all contributions.
  * For new features, documentation must be included.

>   * Once review has occurred, please rebase your PR down to a single commit. This will ensure a nice clean Git history.

Reviewer: Did someone say to include this, we don't need to squash all the commits into one.

Author: I will remove that.

>   

Reviewer 2: Can we also add how the commit messages should be(standards that we follow) and give an example of commit message?

Author: Not sure If we should add that. I agree that there are some common standards that people should follow, and that applies to this repo as well. But I think we should mention guidelines for committing messages only if there are some mandatory rules that need to be followed. For example, `All commit messages should start with Issuenumber`. For Kanister we don't have such rules.
cc @viveksinghggits

Reviewer: I think we should add some guidelines. A set of guidelines where shared in our slack as well. Let's refer them and add a list here.

## Local Development

> Once you are done with your changes, you need to ensure that your changes do not fail in the CI build

Reviewer: Since we have already talked about raising PR don't you think its out of the flow to ask run these things locally now? Shouldn't these things be done before raising PR?

Author: I have made some modifications there

**NOTE:**

> We need to have a local kubernetes cluster to test changes. To spin up a local kubernetes cluster using kind we can run `make start-kind` from the project root directory. Similary to delete the clutser we can run `make stop-kind`

Reviewer: ```suggestion
We need to have a local Kubernetes cluster to test changes. To spin up a local Kubernetes cluster using kind we can run `make start-kind` from the project root directory. Similarly to delete the cluster we can run `make stop-kind`
```

> We need to have a local kubernetes cluster to test changes. To spin up a local kubernetes cluster using kind we can run `make start-kind` from the project root directory. Similary to delete the clutser we can run `make stop-kind`

Reviewer: ```suggestion
We need to have a local kubernetes cluster to test changes. We can use Kind[king-link] to spin up a local kubernetes cluster, `make start-kind` can be run from the project's root directory to create one. Similary to delete the clutser we can run `make stop-kind`
```

```bash
#Run golint command to make sure your code is properly formatted
make golint

#build the project
make build

#Run unit test 
make test

#Run E2E test
#From the project root directory
make integration-test
```

> To test and deploy your changes to a local Kubernetes cluster refer this https://docs.kanister.io/install.html#building-and-deploying-from-source

Reviewer: ```suggestion
To test and deploy your changes to a local Kubernetes cluster [this document](https://docs.kanister.io/install.html#building-and-deploying-from-source) can be referred.
```

## Contributing to documentation
For complete documentation visit https://docs.kanister.io/

Kanister docs are generated using [Sphinx](https://www.sphinx-doc.org/en/master/) and are written in [reStructuredText](https://docutils.sourceforge.io/rst.html). The source `.rst` files are in the Kanister repository under the `/docs` folder.

### Updating documentation
- Modify or add `.rst` file(s) under the `/docs` folder.

- Build Docs locally.
```bash
make docs
```

> - The above command will build the documentation, check for errors and place the final output in the `/docs/_build/html` directory.

Reviewer: do we manually place the files in that directory?

Author: the file is available in that directory after running `make docs`

# Contributing to Kanister

## How Can I Contribute?

### Reporting Bugs
Before creating bug reports, please check the list of issues as you might find out that a similar issue is already created and you don't need to create one. Also when creating a bug report, please include as many details as possible as this information may help us resolve the issue faster.

### Suggesting Enhancements
You can suggest minor improvements to existing functionality and or a completely new feature. For that, we recommend creating an issue at https://github.com/kanisterio/kanister/issues or reaching out to us at our [Slack](https://join.slack.com/t/kanisterio/shared_invite/enQtNzg2MDc4NzA0ODY4LTU1NDU2NDZhYjk3YmE5MWNlZWMwYzk1NjNjOGQ3NjAyMjcxMTIyNTE1YzZlMzgwYmIwNWFkNjU0NGFlMzNjNTk) workspace.

### Your first code contribution
You can start by looking through `good-first-issue` issues. 
  
#### Local Developement

> Once you are done with your changes, you need to ensure that your changes do not fail in the CI build

Reviewer: you need to make sure, that your changes are not breaking anything.

# Contributing to Kanister

## How Can I Contribute?
  ### * Reporting Bugs
  Before creating bug reports, please check the list of issues as you might find out that you don't need to create one. Also when creating a bug report, please include as many details as possible as this information may help us resolve the issue faster.

  ### * Suggesting Enhancements
  You can suggest minor improvements to existing functionality and or a completely new feature. 

  ### * Your first code contribution

>   You can start by looking through `good-first-issue` issues:

Author: I think we should follow this practice to add this label from now on.

### Your first code contribution
You can start by looking through `good-first-issue` issues. 
  
#### Local Developement
Once you are done with your changes, you need to make sure, that your changes are not breaking anything.

**NOTE:**
We need to have a local Kubernetes cluster to test changes. We can use [Kind](https://kind.sigs.k8s.io/) to spin up a local Kubernetes cluster, `make start-kind` can be run from the project's root directory to create one. Similarly to delete the cluster we can run `make stop-kind`

```bash
#Run golint command to make sure your code is properly formatted
make golint

#build the project
make build

#Run unit test 
make test

#Run E2E test
#From the project root directory
make integration-test
```
To test and deploy your changes to a local Kubernetes cluster [this document](https://docs.kanister.io/install.html#building-and-deploying-from-source) can be referred.

> - Push a PR with the changes for review. Please see [Raising PR](#raising-pr)

Reviewer 2: ```suggestion
- Create a pull request with the changes for review. Please refer [Raising PR](#raising-pr)
```

### Contributing to documentation
For complete documentation visit https://docs.kanister.io/

Kanister docs are generated using [Sphinx](https://www.sphinx-doc.org/en/master/) and are written in [reStructuredText](https://docutils.sourceforge.io/rst.html). The source `.rst` files are in the Kanister repository under the `/docs` folder.

### Updating documentation
- Modify or add `.rst` file(s) under the `/docs` folder.

- Build Docs locally.
```bash
make docs
```

- The above command will build the documentation, check for errors and place the final output in the `/docs/_build/html` directory.

- The built docs can be viewed and validated visually by opening `/docs/_build/html/index.html`.

- Push a PR with the changes for review. Please see [Raising PR](#raising-pr)

## Raising PR 
* We follow the [Github Pull Request Model](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) for all contributions.
* For new features, documentation must be included.
* Here are some guidelines that we should follow while writing a commit message:

>   - Use Imperative mood in the commit message. For example, `Add go mod check while init` and not `Added go mod check while init`.

Reviewer 2: ```suggestion
  - Commit msg should be imperative. For example, `Add go mod check while init` and not `Added go mod check while init`.
```

# Contributing to Kanister

## How Can I Contribute?

### Reporting Bugs
Before creating bug reports, please check the list of issues as you might find out that a similar issue is already created and you don't need to create one. Also when creating a bug report, please include as many details as possible as this information may help us resolve the issue faster.

### Suggesting Enhancements

> You can suggest minor improvements to existing functionality and or a completely new feature. For that, we recommend creating an issue at https://github.com/kanisterio/kanister/issues or reaching out to us at our [Slack](https://join.slack.com/t/kanisterio/shared_invite/enQtNzg2MDc4NzA0ODY4LTU1NDU2NDZhYjk3YmE5MWNlZWMwYzk1NjNjOGQ3NjAyMjcxMTIyNTE1YzZlMzgwYmIwNWFkNjU0NGFlMzNjNTk) workspace.

Reviewer 3: Feels a bit redundant since we already have issue templates defined. WDYT?

### Your first code contribution
You can start by looking through `good-first-issue` issues. 
  
#### Local Developement
Once we are done with the changes, we need to make sure, that the changes are not breaking anything and are passing the unit as well as the integration test.

```bash
#Run golint command to make sure your code is properly formatted
make golint

#build the project

> make build

Reviewer: where does `make build` generate the binaries?

> make build

Reviewer: what does this do? Where does it generate the binaries?

Author: and it generates the binaries in `$GOPATH/bin` directory

Author: Resolving this conversation.

# For example ^PostgreSQL$
SHORT_APPS="^PostgreSQL$|^MySQL$|Elasticsearch|^MongoDB$|Maria|^MSSQL$"

# Run E2E test
# From the project root directory
./build/integration-test.sh ${SHORT_APPS}

```
**Note:** We can find the application name from `/pkg/testing/integration_register.go`

Once the integration test passes. Create a pull request with the changes for review. Please refer [Raising PR](#raising-pr)

### Contributing to documentation
For complete documentation visit https://docs.kanister.io/

Kanister docs are generated using [Sphinx](https://www.sphinx-doc.org/en/master/) and are written in [reStructuredText](https://docutils.sourceforge.io/rst.html). The source `.rst` files are in the Kanister repository under the `/docs` folder.

### Updating documentation
- We can modify or add `.rst` file(s) under the `/docs` folder.

- Build Docs locally
```bash
make docs
```

- The above command will build the documentation, check for errors and place the final output in the `/docs/_build/html` directory.

- The built docs can be viewed and validated visually by opening `/docs/_build/html/index.html`.

- Push a PR with the changes for review. Please see [Raising PR](#raising-pr)

> ## Raising PR 

Reviewer 3: Can we bump this section to the top? Most people look at CONTRIBUTING.md because they want to submit a pull request. And it will flow nicely with other sections like "Local Development", "Your First Code Contribution" etc.

* We follow the [Github Pull Request Model](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) for all contributions.
* For new features, documentation must be included.
* Here are some guidelines that we should follow while writing a commit message:
  - Commit message should be imperative. For example, `Add go mod check while init` and not `Added go mod check while init`.
  - Commit message should not start with lower case.
  - Your commit message should not contain any whitespace errors.
  - Remove unnecessary punctuation marks.
  - Use the body to explain what changes you have made and why you made them.

> ## Contacting Developers

Reviewer 3: Can we call this "Getting Help" to convey that non-maintainers might be able to help on slack too?

Author: lets see what other folks have to say on this. cc @viveksinghggits
